assets/app/view/game/stock_market.rb

```diff
@@ -22,47 +22,84 @@ class StockMarket < Snabberb::Component
         gray: '#888888',
       }.freeze
 
+      # All markets
       PAD = 5                                     # between box contents and border
       BORDER = 1
       WIDTH_TOTAL = 50                            # of entire box, including border
+      TOKEN_SIZE = 25
+
+      # 1D markets
+      VERTICAL_TOKEN_PAD = 4                      # vertical space between tokens
+      MIN_NUM_TOKENS = 4                          # guarantee space for this many tokens
+      PRICE_HEIGHT = 20                           # depends on font and size!
+
+      # 2D markets
       HEIGHT_TOTAL = 50
       TOKEN_PAD = 3                               # left/right padding of tokens within box
-      TOKEN_SIZE = 25
       BOX_WIDTH = WIDTH_TOTAL - 2 * BORDER
       LEFT_MARGIN = TOKEN_PAD                     # left edge of leftmost token
       RIGHT_MARGIN = BOX_WIDTH - TOKEN_PAD        # right edge of rightmost token
       LEFT_TOKEN_POS = LEFT_MARGIN
       RIGHT_TOKEN_POS = RIGHT_MARGIN - TOKEN_SIZE # left edge of rightmost token
       MID_TOKEN_POS = (LEFT_TOKEN_POS + RIGHT_TOKEN_POS) / 2
 
-      def render
-        space_style = {
+      def cell_style(box_style, price)
+        style = box_style.merge(backgroundColor: price.color ? COLOR_MAP[price.color] : color_for(:bg2))
+        if price.color == :black
+          style[:color] = 'gainsboro'
+          style[:borderColor] = color_for(:font)
+        end
+        style
+      end
+
+      def grid_1d
+        box_style = {
           position: 'relative',
           display: 'inline-block',
           padding: "#{PAD}px",
-          width: "#{WIDTH_TOTAL - 2 * PAD}px",
-          height: "#{HEIGHT_TOTAL - 2 * PAD}px",
           margin: '0',
           verticalAlign: 'top',
-        }
-
-        box_style = space_style.merge(
           width: "#{WIDTH_TOTAL - 2 * PAD - 2 * BORDER}px",
-          height: "#{HEIGHT_TOTAL - 2 * PAD - 2 * BORDER}px",
           border: "solid #{BORDER}px rgba(0,0,0,0.2)",
           color: color_for(:font2),
-        )
+        }
+
+        tokens_style = {
+          textAlign: 'center',
+          lineHeight: '0',
+        }
+
+        price_style = {
+          fontSize: '100%',
+          textAlign: 'center',
+        }
+
+        max_num_corps = @game.stock_market.market.first.map { |p| p.corporations.size }.push(MIN_NUM_TOKENS).max
```

Comment:
[@game.stock_market.markets[0].map { |p| p.corporations.size }.max, MIN_NUM_TOKENS].max

i found this a bit more readable

Reply:
i'll leave it up to you

Reply:
If you're leaving it up to me, I'll keep it as it is, to have one `max` rather than two.